linux/SystemdMeter.c

```diff
@@ -0,0 +1,328 @@
+static int (*sym_sd_bus_get_property_string)(sd_bus*, const char*, const char*, const char*, const char*, sd_bus_error*, char**);
+static int (*sym_sd_bus_get_property_trivial)(sd_bus*, const char*, const char*, const char*, const char*, sd_bus_error*, char, void*);
+static sd_bus* (*sym_sd_bus_unref)(sd_bus*);
+
+static char* systemState = NULL;
+static unsigned int nFailedUnits = INVALID_VALUE;
+static unsigned int nInstalledJobs = INVALID_VALUE;
+static unsigned int nNames = INVALID_VALUE;
+static unsigned int nJobs = INVALID_VALUE;
+static void* dlopenHandle = NULL;
+static sd_bus* bus = NULL;
+
+static void SystemdMeter_done(ATTR_UNUSED Meter* this) {
+   free(systemState);
+   systemState = NULL;
+
+   if (bus && dlopenHandle) {
+      sym_sd_bus_unref(bus);
+   }
+   bus = NULL;
+
+   if (dlopenHandle) {
+      dlclose(dlopenHandle);
+      dlopenHandle = NULL;
+   }
+}
+
+static int updateViaLib(void) {
+   if (!dlopenHandle) {
+      dlopenHandle = dlopen("libsystemd.so.0", RTLD_LAZY);
+      if (!dlopenHandle)
+         goto dlfailure;
+
+      /* Clear any errors */
+      dlerror();
+
+      *(void **)(&sym_sd_bus_open_system) = dlsym(dlopenHandle, "sd_bus_open_system");
+      if (!sym_sd_bus_open_system || dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_get_property_string) = dlsym(dlopenHandle, "sd_bus_get_property_string");
+      if (!sym_sd_bus_get_property_string || dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_get_property_trivial) = dlsym(dlopenHandle, "sd_bus_get_property_trivial");
+      if (!sym_sd_bus_get_property_trivial || dlerror() != NULL)
+         goto dlfailure;
+
+      *(void **)(&sym_sd_bus_unref) = dlsym(dlopenHandle, "sd_bus_unref");
+      if (!sym_sd_bus_unref || dlerror() != NULL)
+         goto dlfailure;
+   }
+
+   int r;
+
+   /* Connect to the system bus */
+   if (!bus) {
+      r = sym_sd_bus_open_system(&bus);
+      if (r < 0)
+         goto busfailure;
+   }
+
+   r = sym_sd_bus_get_property_string(bus,
+                                      "org.freedesktop.systemd1",           /* service to contact */
+                                      "/org/freedesktop/systemd1",          /* object path */
+                                      "org.freedesktop.systemd1.Manager",   /* interface name */
+                                      "SystemState",                        /* property name */
+                                      NULL,                                 /* object to return error in */
+                                      &systemState);
+   if (r < 0)
+      goto busfailure;
+
+   r = sym_sd_bus_get_property_trivial(bus,
+                                       "org.freedesktop.systemd1",           /* service to contact */
+                                       "/org/freedesktop/systemd1",          /* object path */
+                                       "org.freedesktop.systemd1.Manager",   /* interface name */
```

Comment:
What about having constants for those 3 strings?

Reply:
adopted